Добавлена отправка электронных писем на основе шаблона Razor Pages  + правки шаблона T4

BannikovRuslan opened the pull request:
feat: Добавлена отправка электронных писем на основе шаблона Razor Pages

fix: Поправлен шаблон T4. Удалено ненужное Subject в шаблоне и связанной модели.

BannikovRuslan committed:
Merge remote-tracking branch 'origin/main' into feature-231795-razorpages-template

BannikovRuslan committed:
feat: Добавлена отправка электронных писем на основе шаблона Razor Pages


fix: Поправлен шаблон T4. Удалено ненужное Subject в шаблоне и связанной модели.

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/IEmailSender.cs:
```diff
@@ -32,5 +33,18 @@ public interface IEmailSender
         void SendT4Email(string from, string to, string copyTo, string subject, string body, Dictionary<string, string> bodyAttachments, string fileName, byte[] fileBody);
+
+        /// <summary>
+        /// Отправить письмо, сформированное на основе шаблона Т4.
```
Razor

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/IRazorPageToStringRenderer.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading.Tasks;
```
using внутри namespace

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -18,9 +20,38 @@ public class MailKitEmailService : IEmailSender
+        /// <param name="bodyAttachments">Прикреляемые изображения для отображения в теле письма.</param>
+        /// <param name="fileName">Имя прикрепляемого файла.</param>
+        /// <param name="fileBody">Содержимое прикрепляемого файла.</param>
+        public async Task SendRazorPagesEmail(string from, string to, string copyTo, string subject, string body, Dictionary<string, string> bodyAttachments, string fileName, byte[] fileBody)
```
А почему с Razor асинхронная операция, а с T4 нет. В примерах существующих проектов также ?

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -38,11 +69,10 @@ public void SendT4Email(string from, string to, string copyTo, string subject, s
             };
             string messageBody = mailCommonRu.TransformText();
-            SendEmail(from, to, copyTo, subject, messageBody, bodyAttachments, fileName, fileBody);
+            this.SendEmail(from, to, copyTo, subject, messageBody, bodyAttachments, fileName, fileBody);
```
this тут (и везде ниже) не надо. Валентин должен был там прикрутить правила с которыми не будет на это ругаться, надо main втащить

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/RazorViewToStringRenderer.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public RazorViewToStringRenderer(IRazorViewEngine viewEngine, ITempDataProvider tempDataProvider, IServiceProvider serviceProvider)
+        {
+            this._viewEngine = viewEngine;
```
Мы используем this, поэтому черточки в именах локальных переменных не нужно

А что такое ITempDataProvider и IServiceProvider , это чисто Razorовское что-то ?

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Startup.cs:
```diff
@@ -1,6 +1,10 @@
+    using System.ComponentModel;
+    using System.ComponentModel.Design;
+    using System.Text;
+    using DocumentFormat.OpenXml.Office2016.Drawing.ChartDrawing;
```
А это для чего такая интересная штука ?

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Startup.cs:
```diff
@@ -21,6 +30,8 @@
     using NewPlatform.Flexberry.ORM.ODataServiceCore.Common.Exceptions;
     using NewPlatform.Flexberry.Services;
     using Unity;
+    using Unity.Microsoft.DependencyInjection;
```
И это для чего добавилось ?
Проверить, если есть неиспользуемые using то их удалить надо

bratchikov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Startup.cs:
```diff
@@ -85,6 +96,8 @@ public void ConfigureServices(IServiceCollection services)
                 .AddHealthChecks()
                 .AddNpgSql(connStr);
+
+            services.AddRazorPages();
```
Разве нам не просто рендерер от Razor нужен? Посмотри как сделано в примерах у Антона М..

bratchikov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Startup.cs:
```diff
@@ -120,6 +133,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                 endpoints.MapControllers();
                 endpoints.MapHealthChecks("/health");
                 endpoints.MapHub<SignalRHub>("/SignalRTest");
+                endpoints.MapRazorPages();
```
У нас же не будет Razor pages, мы просто рендерим один шаблончик в строку и всё.

bratchikov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Templates/MailTemplates/RazorPages/RazorPagesMailTemplate.cshtml.cs:
```diff
@@ -0,0 +1,12 @@
+    public class RazorPagesMailTemplateModel : PageModel
+    {
+        /// <summary>
+        /// ����� ��������� � ������� HTML.
```
Перепроверь, что у тебя все новые файлы в этом PR сохранены в UTF-8.

bratchikov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Templates/MailTemplates/T4MailTemplate.tt:
```diff
@@ -5,7 +5,7 @@
 <html xmlns="http://www.w3.org/1999/xhtml">
 <head>
-    <title><#= this.Subject #> <#= this.Subject #></title>
+    <title></title>
```
Надо хоть какой-то заголовок зашить в шаблон.

BannikovRuslan committed:
fix: внесены рекомендуемые правки.

AntonIV8 commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -38,7 +70,7 @@ public void SendT4Email(string from, string to, string copyTo, string subject, s
             T4MailTemplate mailCommonRu = new T4MailTemplate()
             {
-                Subject = subject,
+                HeadTitle = "Письмо на базе шаблона T4",
```
Такого в сервисе не должно быть. Это хардкод.
HeadTitle я считаю нужно убрать (и в остальных связанных классах тоже), т.к мы его никак не можем поменять в этом сервисе - нет переменной к нему.
А сам текст "Письмо на базе шаблона T4" должен быть зашит в шаблон.
Это к шаблону Razor тоже относится.
Если хочется включить кастомизацию заголовка (хотя по мне так не надо) то нужно включать его везде в параметры

BannikovRuslan committed:
fix: заголовок занесен в шаблон и убран из модели, файлы шаблона Т4 перемещены в отдельную папку

VChurekov commented:
> SonarCloud Quality Gate failed.    [![Quality Gate failed](https://camo.githubusercontent.com/4ea51c1f64ee3746f631653a02ab678ca6a3efb5f5cb474402faed2e3dcf90b5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f6661696c65642d313670782e706e67)](https://sonarcloud.io/dashboard?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38)
> 
> [![Bug](https://camo.githubusercontent.com/4c6102327f5a954f9c8acaf2e2714183157a9e41717b371b2cd585cf25057310/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=BUG) [![C](https://camo.githubusercontent.com/57cf2cd81158730ed1dd8be1e93a41a2feb8b35dd75e5601402ba5f961f6ec18/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f432d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=BUG) [10 Bugs](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=BUG) [![Vulnerability](https://camo.githubusercontent.com/3ba1ee49636ffc3427e38649a9f8a65ee392f28e8a662fcf96ce24cefbb520e9/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=VULNERABILITY) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=VULNERABILITY) [![Security Hotspot](https://camo.githubusercontent.com/fb735cbe76f8d5e1679c76ce83b740ceb1eaf62de4f7bf88623dc9953261aff7/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=SECURITY_HOTSPOT) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/security_hotspots?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=SECURITY_HOTSPOT) [![Code Smell](https://camo.githubusercontent.com/8fe18b2dfb6f7d4e44582f281b29f617eb5ae07c248d2002ca586e91da219212/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=CODE_SMELL) [![A](https://camo.githubusercontent.com/1cba125a897d7fa47033a3b3b2be2bbee680d34d4f004a215564659b853fb201/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412d313670782e706e67)](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=CODE_SMELL) [54 Code Smells](https://sonarcloud.io/project/issues?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&resolved=false&types=CODE_SMELL)
> 
> [![No Coverage information](https://camo.githubusercontent.com/86785eaa572e061f5ecb3f8bf1cbb7b8eca44925083f92d067805b4362cead0c/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f4e6f436f766572616765496e666f2d313670782e706e67)](https://sonarcloud.io/component_measures?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38) No Coverage information [![0.0%](https://camo.githubusercontent.com/8047c63e1f9ed03f63001e1eadce4676bade3e0f83ec690a9c625287796248a6/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332d313670782e706e67)](https://sonarcloud.io/component_measures?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Flexberry_Flexberry.AsyncOpenXmlReports.Sample&pullRequest=38&metric=new_duplicated_lines_density&view=list)

Еще обратите внимание чтобы сонар не ругался на Bugs и Code Smells.
Сейчас 10 ошибок и 54 места... где код воняет? Никогда не понимал как это перевести...

BannikovRuslan committed:
fix: исправлены баги по sonar + комментарий метода

BannikovRuslan committed:
fix: sonarcloud Code Smells

BannikovRuslan committed:
fix: sonarcloud Code Smells

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/IEmailSender.cs:
```diff
@@ -45,6 +45,7 @@ public interface IEmailSender
         /// <param name="bodyAttachments">Прикреляемые изображения для отображения в теле письма.</param>
         /// <param name="fileName">Имя прикрепляемого файла.</param>
         /// <param name="fileBody">Содержимое прикрепляемого файла.</param>
+        /// <returns>Ничего.</returns>
```
Ничего? Может тут хотя бы указать просто Task. Если ничего то надо переделать на void, но я так понимаю что модель взаимодействия подразумевает именно Task.

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -42,15 +47,16 @@ public MailKitEmailService(EmailOptions emailOptions, IRazorViewToStringRenderer
         /// <param name="bodyAttachments">Прикреляемые изображения для отображения в теле письма.</param>
         /// <param name="fileName">Имя прикрепляемого файла.</param>
         /// <param name="fileBody">Содержимое прикрепляемого файла.</param>
+        /// <returns>A <see cref="Task"/> Ничего.</returns>
```
Может не надо писать слово "Ничего"?

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -42,15 +47,16 @@ public MailKitEmailService(EmailOptions emailOptions, IRazorViewToStringRenderer
 
             string razorPage = "/Templates/MailTemplates/RazorPages/RazorPagesMailTemplate.cshtml";
-            string messageBody = await razorPageToStringRenderer.RenderViewToStringAsync(razorPage, mailModel);
+            string messageBody = await razorPageToStringRenderer.RenderViewToStringAsync(razorPage, mailModel).ConfigureAwait(true);
```
.ConfigureAwait(false)

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -123,6 +133,18 @@ private SmtpClient GetSmtpClient()
+        /// <param name="fileName">Имя прикрепляемого файла.</param>
+        /// <param name="fileBody">Содержимое прикрепляемого файла.</param>
+        /// <returns>Сформированное сообщение.</returns>
         private MimeMessage GetMessage(string from, string to, string copyTo, string subject, string body, Dictionary<string, string> bodyAttachments, string fileName, byte[] fileBody)
```
Можно ли переименовать параметры from и to? Они вроде как пересекаются с какими-то служебными словами...

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/MailKitEmailService.cs:
```diff
@@ -42,15 +47,16 @@ public MailKitEmailService(EmailOptions emailOptions, IRazorViewToStringRenderer
         /// <param name="fileName">Имя прикрепляемого файла.</param>
         /// <param name="fileBody">Содержимое прикрепляемого файла.</param>
+        /// <returns>A <see cref="Task"/> Ничего.</returns>
         public async Task SendRazorPagesEmail(string from, string to, string copyTo, string subject, string body, Dictionary<string, string> bodyAttachments, string fileName, byte[] fileBody)
```
Можно ли переименовать параметры from и to? Они вроде как пересекаются с какими-то служебными словами...

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/RazorViewToStringRenderer.cs:
```diff
@@ -14,19 +14,35 @@
+        /// </summary>
+        /// <param name="viewEngine">Razor View Engine.</param>
+        /// <param name="tempDataProvider">Temp Data Provider.</param>
+        /// <param name="serviceProvider">Service Provider.</param>
```
Если везде описание на русском, то и тут надо на русском сделать. Да и эти описания по сути повторяют наименования переменных.

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Services/RazorViewToStringRenderer.cs:
```diff
@@ -0,0 +1,98 @@
+                output,
+                new HtmlHelperOptions());
+
+            await view.RenderAsync(viewContext).ConfigureAwait(true);
```
ConfigureAwait(false)

VChurekov commented on src/AsyncOpenXmlReportsSample/ODataBackend/Templates/MailTemplates/T4/T4MailTemplate.cs:
```diff
@@ -28,43 +28,30 @@ public partial class T4MailTemplate : T4MailTemplateBase
-                    "          <td bgcolor=\"#ffffff\" align=\"left\" style=\"padding: 20px 30px 40px 30px" +
-                    "; color: #666666; font-family: \'Lato\', Helvetica, Arial, sans-serif; font-size: " +
-                    "18px; line-height: 18px;\">\r\n                      ");
+            this.Write("<html xmlns=\"http://www.w3.org/1999/xhtml\">\r\n<head>\r\n    <title>Письмо на базе ша" +
```
О боже! А нельзя эту строку в файл вынести и вычитать потом из файла как строку? А потом уже использовать строку в этом методе...

BannikovRuslan replied:
Это генерируется автоматически.

VChurekov commented on src/AsyncOpenXmlReportsSample/Tests/MailKitEmailServiceTests.cs:
```diff
@@ -0,0 +1,109 @@
+        private string From { get; set; } = "bannikov.ru@inbox.ru";
+
+        private string Password { get; set; } = "7qtnTN8a00fWmP1mQWv3";
+
```
Эти данные должны быть в файле конфигурации. Более того там уже есть данные тестового ящика, не надо свои личные данные тут светить...

VChurekov commented on src/AsyncOpenXmlReportsSample/Tests/MailKitEmailServiceTests.cs:
```diff
@@ -0,0 +1,109 @@
+        {
+            return new EmailOptions
+            {
+                Host = "smtp.mail.ru",
```
Взять данные из конфигурации.

VChurekov commented on src/AsyncOpenXmlReportsSample/Tests/MailKitEmailServiceTests.cs:
```diff
@@ -0,0 +1,109 @@
+            return new EmailOptions
+            {
+                Host = "smtp.mail.ru",
+                Port = 465,
```
Взять данные из конфигурации.

BannikovRuslan committed:
fix: исправления по замечаниям Валентина

VChurekov commented:
Пробежался по замечаниям. Закрыл те по которым увидел исправления. Осталось два. По ним, если исправления не нужны, то лучше написать комментарий и закрыть.
